tsdb/exemplar.go

```diff
@@ -250,10 +250,29 @@ func (ce *CircularExemplarStorage) validateExemplar(key []byte, e exemplar.Exemp
 	}
 
 	if e.Ts <= ce.exemplars[idx.newest].exemplar.Ts {
-		if appended {
-			ce.metrics.outOfOrderExemplars.Inc()
+		// Check for duplicates in the stored exemplars for this series.
+		// NB these are expected, and appending them is a no-op.
+		// This is separate from the earlier check because the earlier one is
+		// expected to be triggered more often and we don't want to go through
+		// the exemplars for this series unnecessarily.
+		// This triggers even when the timestamp is equal to the newest exemplar
+		// in case there are multiple exemplars with the same timestamp like for
+		// native histograms. However this means that we may repeat the comparison
+		// for the newest exemplar.
+		for i := idx.oldest; i != noExemplar; i = ce.exemplars[i].next {
```

Comment:
After discussion IRL:

If there was no computational cost we'd do this, however there is and this code runs under lock as well, so we'd like to avoid doing it.

Looking at the check , what it really does is distinguish between the case where an "old" exemplar is either out of order or a duplicate. But this is expensive to do when we can have multiple exemplars per sample (for whatever reason, it could be other than native histograms).

Use case1:  incoming exemplars [ts=1, ts=2, ts=3], then incoming exemplars: [ts=1,ts=2,ts=4]. The first two will probably be duplicates and ts=4  a new exemplar.
Use case2: incoming exemplars [ts=10, ts=11, ts=13], then incoming exemplars [ts=1, ts=2, ts=3]. The whole thing seems like out of order.

Suggested solution: approximate the result by only looking at the last exemplar in the incoming list of **sorted** exemplars. This means having a new validate exemplar function that takes a slice as input.

If the newest exemplar in the new list is out of order, then consider all out of order. Otherwise assume that the list contains duplicates (don't check, don't report as ooo) and new exemplars (append).

Reply:
A possible still suboptimal , but slightly better fix https://github.com/zenador/prometheus/pull/9